torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -2111,10 +2111,28 @@ Tensor symeig_backward(const std::vector<torch::autograd::Variable> &grads, cons
 
   auto vh = v.conj().transpose(-2, -1);
 
+  // Based on a little experiment, min_threshold should be
+  //   "max_eigenvalue * matrix_size * machine_eps"
+  // See: https://github.com/pytorch/pytorch/pull/50942
+  // However, as getting max_eigenvalue would need synchronization, I just take
+  //   a guess that the max eigenvalues would typically be less than 10
+  // Eps is still taking double and single eps value, because I don't know
+  //   if it would produce a lot of backward incompatibility due to large
+  //   eps for half-precision floating point
+  double eps = (lambda.dtype() == at::kDouble) ? 2.220446049250313e-16 : 1.1920928955078125e-07;
+  double min_threshold = 10 * lambda.size(-1) * eps;
+
   Tensor result;
   if (gv.defined()) {
       Tensor F = lambda.unsqueeze(-2) - lambda.unsqueeze(-1);
-      F.diagonal(/*offset=*/0, /*dim1=*/-2, /*dim2=*/-1).fill_(INFINITY);
+
+      // simple modification to make it numerically stable if there are
+      // repeated eigenvalues
+      // see https://github.com/pytorch/pytorch/issues/47599
+      // and https://arxiv.org/abs/2011.04366
+      Tensor idx = F.abs() <= min_threshold;
```

Comment:
How's this inequality gonna perform in half precision? If `min_threshold` is first converted to fp16, it's probably going to be zero, but given that it's less_or_equal, it's probably fine?

Reply:
Yes, for half precision if `min_threshold` becomes 0, it has the same effect as the original algorithm.

Reply:
Maybe worth adding an explicit comment about this in the code, so that next reader won't have to think about correctness for fp16 wrt to eps becoming zero?

Reply:
Good suggestion. Did it in the newest commit.

Reply:
AFAIK `torch.symeig` and `torch.linalg.eigh` do not support fp16 input. So `lambda` and `F` shouldn't have fp16 dtype for now.

Reply:
Yes, I confirm, floats and doubles only.